presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerClusterStateProvider.java

```diff
@@ -0,0 +1,481 @@
+        this.completedQueryExpirationTimeout = requireNonNull(completedQueryExpirationTimeout, "completedQueryExpirationTimeout is null");
+        // Memoized suppliers take in a time unit > 0
+        requireNonNull(memoryPoolInfoRefreshDuration, "memoryPoolInfoRefreshDuration is null");
+        if (memoryPoolInfoRefreshDuration.toMillis() > 0) {
+            this.clusterMemoryPoolInfosSupplier = Suppliers.memoizeWithExpiration(this::getClusterMemoryPoolInfoInternal, memoryPoolInfoRefreshDuration.toMillis(), MILLISECONDS);
+        }
+        else {
+            this.clusterMemoryPoolInfosSupplier = this::getClusterMemoryPoolInfoInternal;
+        }
+        this.isReservedPoolEnabled = isReservedPoolEnabled;
+
+        requireNonNull(scheduledExecutorService, "scheduledExecutorService is null");
+        scheduledExecutorService.scheduleAtFixedRate(() -> {
+            for (CoordinatorQueriesState coordinatorQueriesState : ImmutableList.copyOf(nodeQueryStates.values())) {
+                coordinatorQueriesState.purgeExpiredQueries();
+            }
+        }, 100, 100, MILLISECONDS);
+
+        scheduledExecutorService.scheduleAtFixedRate(() -> {
+            for (Map.Entry<String, InternalNodeState> nodeEntry : ImmutableList.copyOf(nodeStatuses.entrySet())) {
+                if ((System.currentTimeMillis() - nodeEntry.getValue().getLastHeartbeatInMillis()) > nodeStatusTimeout.toMillis()) {
+                    nodeStatuses.remove(nodeEntry.getKey());
+                }
+            }
+        }, 100, 100, MILLISECONDS);
+    }
+
+    public void registerQueryHeartbeat(String nodeId, BasicQueryInfo basicQueryInfo, long sequenceId)
+    {
+        requireNonNull(nodeId, "nodeId is null");
+        requireNonNull(basicQueryInfo, "basicQueryInfo is null");
+        Stream<InternalNode> activeOrShuttingDownCoordinators = concat(internalNodeManager.getCoordinators().stream(),
+                internalNodeManager.getShuttingDownCoordinator().stream());
+        checkArgument(
+                activeOrShuttingDownCoordinators.anyMatch(i -> nodeId.equals(i.getNodeIdentifier())),
+                "%s is not a coordinator (coordinators: %s)",
+                nodeId,
+                internalNodeManager.getCoordinators().stream().collect(toImmutableSet()));
+        CoordinatorQueriesState state = nodeQueryStates.computeIfAbsent(nodeId, identifier -> new CoordinatorQueriesState(
+                nodeId,
+                maxCompletedQueries,
+                queryExpirationTimeout.toMillis(),
+                completedQueryExpirationTimeout.toMillis()));
+        state.addOrUpdateQuery(basicQueryInfo, sequenceId);
+    }
+
+    public void registerNodeHeartbeat(NodeStatus nodeStatus)
+    {
+        requireNonNull(nodeStatus, "nodeStatus is null");
+        InternalNodeState nodeState = nodeStatuses.get(nodeStatus.getNodeId());
+        if (nodeState == null) {
+            nodeStatuses.put(nodeStatus.getNodeId(), new InternalNodeState(nodeStatus));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **164, 165, and 166**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).